sql: fix temporary schema cleaner to use transaction correctly

ajwerner opened the pull request:
This thing before was buggy in some way I don't really understand.

Release justification: Fixes tests and improves code.

Release note: None

ajwerner committed:
sql: fix temporary schema cleaner to use transaction correctly

This thing before was buggy in some way I don't really understand.

Release justification: Fixes tests and improves code.

Release note: None

Commit:
sql/*,descs: clean up CollectionFactory.Txn

It now can now share all the pre-commit logic from the SQL package.

Release justification: non-functional change which improves code

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86301)
<!-- Reviewable:end -->

ajwerner committed:
kv: remove PrepareRetryableError from the (*kv.Txn) API

This error was never what anybody wanted. It has the odd side-effect
of leaving the transaction usable -- and, worse, the (*kv.Txn).exec loop would
retry but would not restart or replace the transaction if such an error were
returned.

Release justification: important correctness change as part of making the
SQL-over-HTTP APIs work for index recommendations.

Release note: None

ajwerner committed:
kv: update a comment to reflect reality

Release justification: Does not change production code, only comments.

Release note: None

ajwerner committed:
fixup! kv: remove PrepareRetryableError from the (*kv.Txn) API

ajwerner committed:
sql: handle jobs properly in a transaction

We need to track the jobs records and we need to run the jobs.

Release note: None

ajwerner committed:
server: proper transaction state management in sql-over-http

We need to construct the internal executor in the context of the transaction
so that we can make sure that its side-effects are properly managed. Without
this change, we'd be throwing away all of the extraTxnState between each
statement. We'd fail to create the jobs (which we defer to the end of the
transaction), and we'd fail to run those jobs and check for errors. We'd
also fail to validate the two-version invariant or wait for one version.

Fixes #86332

Release justification: Fixes critical bugs in new functionality.

Release note: None

ZhouXing19 commented on pkg/sql/internal.go:
```diff
@@ -147,6 +147,7 @@ func newInternalExecutorWithTxn(
 			txn:                    txn,
 			descCollection:         descCol,
+			jobs:                   new(jobsCollection),
 			schemaChangeJobRecords: schemaChangeJobRecords,
```
This is a confusing part that I introduced 😓 We should have had `schemaChangeJobRecords: make(map[descpb.ID]*jobs.Record)` here and remove it from the parameter list. We should also have added a comment that except descriptor collection, all metadata (such as jobs) is this internal executor's own set, thus their lifecycle is managed by the internal executor itself, and independent of the caller.

ajwerner replied:
The truth of the matter is that now that we're here, it seems like the `InternalExecutorFactory` should also be constructing the `descs.Collection` and should own its lifecycle. There's no reason for `descs.CollectionFactory.Txn` anymore. Honestly, that's sort of beautiful. We'll leave that for later. I moved the map construction for now.

ZhouXing19 commented on pkg/sql/catalog/descs/txn.go:
Not a blocker, but why are some of the checks here no longer needed?

ZhouXing19 reviewed (commented):
Thanks for making these changes! Took a rough look at the internal-executor-related parts and left nit comments

ajwerner commented:
superseded by #86427
